### DISCUSSION TITLE
example of how to transform selected text?

### DISCUSSION THREAD

**Discussion by ExpandingMan** [L3MON4D3/LuaSnip#884] [Q&A] [answered]
I know this is a common use case, and probably everyone knows how to do it, but I can't for the life of me seem to figure it out.  I assume I have to set a binding in visual mode but I don't know what to set it to.  I'm also unsure if I'll have to involve my auto complete for the sake of bringing up options.

If it helps for the sake of an example, the luasnip portion of my config is [here](https://gitlab.com/ExpandingMan/neovimconfig/-/blob/master/lua/etc/snippets.lua).

**Comment by L3MON4D3** [comment]
That's the one key-binding we take care of :sweat_smile:
Check `store_selection_keys` [here](https://github.com/L3MON4D3/LuaSnip/blob/master/DOC.md#config-options)
(essentially, do `ls.setup({store_selection_keys = "<Tab>"})`, for example)
(this is done differently from other mappings because the selection can't yet be cut with a lua-function-call, and I'd like to keep the implementation opaque)

**Comment by ExpandingMan** [comment]
Sorry, I don't think I'm understanding how to use this... so it sounds like that binding will set a variable for luasnip `TM_SELECTED_TEXT`... but then I need to set another binding to insert snippets?  Again, this seems tricky since I suppose I'd have to do it with my completion engine (`cmp` in my case)...

I don't suppose you can provide a full example of how to set a key binding to show suggestions when I have highlighted text in visual mode?

Btw, I can't seem to get it to set anything as `TM_SELECTED_TEXT` though I might be overriding the binding somewhere.

**Reply by L3MON4D3** [reply]
You can see one snippet that makes use of `TM_SELECTED_TEXT` [here](https://github.com/L3MON4D3/LuaSnip/blob/b4bc24c4925aeb05fd47d2ee9b24b7f73f5d7e32/Examples/snippets.lua#LL394C4-L394C4).
You'd have to
1. Select some text in visual mode
2. Hit whatever you set as store_selection_keys
3. Expand that snippet

And it should show the transformed text

**Reply by ExpandingMan** [reply]
Hm... ok, this is working as expected, I just have to figure out how to get it to work with `cmp`... I think what I want is to hit `tab` and then it brings up a popup of snippet options... if by any chance you know of an example that does this with `cmp` that would be awesome, otherwise it's something I'll have to tinker with.  Thanks!

**Reply by ExpandingMan** [reply]
On second thought, I guess `cmp` isn't necessarily useful here at all, I just have to figure out how I'm going to show a popup with snippet options that include `TM_SELECTED_TEXT`.

**Reply by L3MON4D3** [reply, marked as answer]
Oh okay that's a bit more complicated, aka not really supported natively at all
There's cmp_luasnip for listing all snippets, but it doesn't do filtering
Why not just use regular triggers though? Do you want some kind of reminder what snippets are available?

**Reply by ExpandingMan** [reply]
Ah, I think I understand now, I was failing to understand that there's already a pretty good way to use this as is.

So I guess the procedure is to type, then hit e.g. `bve` to select the last word, then with this binding I can hit `<tab>` (which helpfully consumes the selected text) and then start typing to get suggestions... that's pretty good, it just took me a while to get it.

Thanks for your help!

**Reply by L3MON4D3** [reply]
Yes, that's the canonical way of doing it :+1: 
Anytime :)
